blockchain/state_manager/src/lib.rs

```diff
@@ -122,6 +138,83 @@ where
         Ok((state_root, rect_root))
     }
 
+    pub async fn tipset_state(&self, tipset: &Tipset) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        span!("tipset_state", {
+            trace!("tipset {:?}", tipset.cids());
+            // if exists in cache return
+            if let Some(cid_pair) = self.cache.read().await.get(&tipset.key()) {
+                return Ok(cid_pair.clone());
+            }
+
+            if tipset.len() == 0 {
+                // NB: This is here because the process that executes blocks requires that the
+                // block miner reference a valid miner in the state tree. Unless we create some
+                // magical genesis miner, this won't work properly, so we short circuit here
+                // This avoids the question of 'who gets paid the genesis block reward'
+                let message_receipts = tipset
+                    .blocks()
+                    .first()
+                    .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+                let cid_pair = (
+                    tipset.parent_state().clone(),
+                    message_receipts.message_receipts().clone(),
+                );
+                self.cache
+                    .write()
+                    .await
+                    .insert(tipset.key().clone(), cid_pair.clone());
+                return Ok(cid_pair);
+            }
+
+            let block_headers = tipset.blocks();
+            // generic constants are not implemented yet this is a lowcost method for now
+            let cid_pair = self.compute_tipset_state(&block_headers)?;
+            self.cache
+                .write()
+                .await
+                .insert(tipset.key().clone(), cid_pair.clone());
+            Ok(cid_pair)
+        })
+    }
+
+    pub fn compute_tipset_state<'a>(
+        &'a self,
+        blocks: &[BlockHeader],
+    ) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        span!("compute_tipset_state", {
+            let check_for_duplicates = |s: &BlockHeader| {
+                blocks
+                    .iter()
+                    .filter(|val| val.miner_address() == s.miner_address())
+                    .take(2)
+                    .count()
+            };
+            if blocks.iter().any(|s| check_for_duplicates(s) > 1) {
+                // Duplicate Miner found
+                return Err(Box::new(Error::Other(
+                    "Could not get message receipts".to_string(),
+                )));
+            }
```

Comment:
as a general rule I usually avoid indexes where I can when using iterators, I think following high order functions is so easier cause you can essentially tell what it is doing without looking at any index arithmetic. I think from a performance perspective it is not as bad because filter is lazy, take is shortcircuiting and count will iter on the take iterator

Reply:
So the argument for this is readability or are you saying you don't like indexing wrt performance? That's fine I mean it's such a trivial cost to have duplicate address comparisons, just I guess myself personally the code was confusing to read what it was actually doing.

Reply:
What is funny is actually find the suggestion much harder to read haha I think it is because I avoid indexes most of the time, I find high order functions much easier to follow since they denote functional flow